File: ICPC/Practica_para_Gran_Premio_de_Mexico_2/H_Robotic_Skills.py
import sys

input = sys.stdin.read
d = input().split()

n = int(d[0])
a = []
idx = 1
for i in range(n):
    r = []
    for j in range(n):
        r.append(int(d[idx]))
        idx += 1
    a.append(r)

pos = [(i, j) for i in range(n) for j in range(n)]

pos.sort(key=lambda x: a[x[0]][x[1]])

x = [0] * (n * n)
y = [0] * (n * n)
res = 0

for i, j in pos:
    antX = x[i]
    antY = y[j]

    y[j] = max(y[j], antX + 1)
    x[i] = max(x[i], antY + 1)

    # print(f"({i}, {j}) -> {x[i]}")

    res = max(res, x[i], y[j])

print(res)

File: ICPC/Practica_para_Gran_Premio_de_Mexico_2/H_Robotic_Skills.cpp
#include<bits/stdc++.h>
using namespace std;

#define sz(a)   (int) a.size()
#define all(a)  a.begin(), a.end()

typedef long long ll;
typedef pair<int, int> pi;

int main () {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n;
    cin >> n;
    vector<vector<int>> a(n, vector<int>(n));
    vector<pi> pos;
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < n; ++j) {
            cin >> a[i][j];
            pos.emplace_back(i, j);
        }
    }

    sort(all(pos), [&](auto x, auto y) {
        return a[x.first][x.second] < a[y.first][y.second];
    });

    vector<int> x(n * n, 0);
    vector<int> y(n * n, 0);
    int res = 0;

    for (auto [i, j]: pos) {
        int antX = x[i]; 
        int antY = y[j]; 

        y[j] = max(y[j], antX + 1); 
        x[i] = max(x[i], antY + 1); 
        
        res = max({res, x[i], y[j]});
    }

    // Python >>>>> c++
    cout << res << "\n";

    return 0;
}